File: src/WireUnpacker.cpp
/**
 * Modified based on https://github.com/gutierrezps/ESP32_I2C_Slave
 */

#include "WireUnpacker.h"

#include "WireCrc.h"

WireUnpacker::WireUnpacker(uint32_t bufferSize)
    : bufferSize_(bufferSize) {
    buffer_ = (uint8_t *)malloc(sizeof(uint8_t) * bufferSize_);
    if (buffer_ == nullptr) {
        bufferSize_ = 0;  // handle allocation error
    }
    reset();
}

WireUnpacker::~WireUnpacker() {
    free(buffer_);
}

size_t WireUnpacker::write(uint8_t data) {
    if (totalLength_ >= bufferSize_ || hasError()) {
        return 0;
    }

    if (!isPacketOpen_) {
        // enable writing only if buffer is empty
        if (totalLength_ == 0 && data == frameStart_) {
            isPacketOpen_ = true;
            ++totalLength_;
            return 1;
        }
        return 0;
    }

    // first byte after start is packet length
    if (expectedLength_ == 0 || numBufferLength_ > 0) {
        if (numBufferLength_ == 0) {
            numBufferLength_ = data;
            if (numBufferLength_ == 1) {
                ignoreLength_ = 5;
            } else if (numBufferLength_ == 2) {
                ignoreLength_ = 6;
            } else {
                ignoreLength_ = 7;
            }
            ++totalLength_;
            return 1;
        }
        expectedLength_ = (expectedLength_ << 8) + data;
        --numBufferLength_;
        if (numBufferLength_ > 0) {
            ++totalLength_;
            return 1;
        }

        if (expectedLength_ > bufferSize_) {
            isPacketOpen_ = false;
            lastError_ = INVALID_LENGTH;
            return 0;
        }

        ++totalLength_;
        return 1;
    }

    // if end byte index wasn't reached
    if (totalLength_ < (expectedLength_ - 1)) {
        buffer_[index_] = data;
        ++index_;
        ++totalLength_;
        return 1;
    }

    isPacketOpen_ = false;

    // add end byte
    ++totalLength_;

    if (data != frameEnd_) {
        lastError_ = INVALID_LENGTH;
        return 0;
    }

    // ignore start, length, crc and end bytes
    payloadLength_ = totalLength_ - ignoreLength_;

    WireCrc crc8;
    uint8_t crc = crc8.update(buffer_, payloadLength_);

    if (crc != buffer_[index_ - 1]) {
        lastError_ = INVALID_CRC;
        return 0;
    }

    index_ = 0;
    return 1;
}

size_t WireUnpacker::write(const uint8_t *data, size_t quantity) {
    for (size_t i = 0; i < quantity; ++i) {
        if (!write(data[i])) {
            return i;
        }
    }
    return quantity;
}

size_t WireUnpacker::available(void) {
    if (isPacketOpen_) return 0;

    return payloadLength_ - index_;
}

int WireUnpacker::read(void) {
    int value = -1;
    if (!isPacketOpen_ && index_ < payloadLength_) {
        value = buffer_[index_];
        ++index_;
    }
    return value;
}

void WireUnpacker::reset() {
    index_ = 0;
    totalLength_ = 0;
    ignoreLength_ = 0;
    numBufferLength_ = 0;
    payloadLength_ = 0;
    expectedLength_ = 0;
    expectedCrc_ = 0;
    isPacketOpen_ = false;
    lastError_ = WireUnpacker::NONE;
}

File: i2c_for_esp32/wireunpacker.py
from i2c_for_esp32.wirecrc import WireCrc


class WireUnpacker(object):

    NONE = 0
    INVALID_CRC = 1
    INVALID_LENGTH = 2

    def __init__(self, buffer_size):
        self.buffer_size = buffer_size

        self.frameStart = 0x02
        self.frameEnd = 0x04
        self.buffer = [0] * self.buffer_size
        self.index = 0
        self.totalLength = 0
        self.payloadLength = 0
        self.isPacketOpen = False
        self.expectedLength = 0
        self.expectedCrc = 0
        self.lastError = self.NONE
        self.reset()

    def hasError(self):
        return self.lastError != self.NONE

    def write(self, data):
        if self.totalLength >= self.buffer_size or self.hasError():
            return 0

        if not self.isPacketOpen:
            if self.totalLength == 0 and data == self.frameStart:
                self.isPacketOpen = True
                self.totalLength += 1
                return 1
            return 0

        if self.expectedLength == 0 or self.numBufferLength > 0:
            if self.numBufferLength == 0:
                self.numBufferLength = data
                if self.numBufferLength == 1:
                    self.ignoreLength = 5
                elif self.numBufferLength == 2:
                    self.ignoreLength = 6
                else:
                    self.ignoreLength = 7
                self.totalLength += 1
                return 1

            self.expectedLength = (self.expectedLength << 8) + data
            self.numBufferLength -= 1
            if self.numBufferLength > 0:
                self.totalLength += 1
                return 1
            if self.expectedLength > self.buffer_size:
                self.isPacketOpen = False
                self.lastError = self.INVALID_LENGTH
                return 0

            self.totalLength += 1
            return 1

        if self.totalLength < (self.expectedLength - 1):
            self.buffer[self.index] = data
            self.index += 1
            self.totalLength += 1
            return 1

        self.isPacketOpen = False
        self.totalLength += 1

        if data != self.frameEnd:
            self.lastError = self.INVALID_LENGTH
            return 0

        self.payloadLength = self.totalLength - self.ignoreLength

        crc8 = WireCrc()
        crc = crc8.update(self.buffer, self.payloadLength)

        if crc != self.buffer[self.index - 1]:
            self.lastError = self.INVALID_CRC
            return 0

        self.index = 0
        return 1

    def write_data_list(self, byte_data_list):
        if byte_data_list[0] != self.frameStart \
           or self.hasError() or self.totalLength >= self.buffer_size:
            return 0

        self.isPacketOpen = True
        self.totalLength += 1

        numBufferLength = byte_data_list[1]
        if numBufferLength == 1:
            self.ignoreLength = 5
        elif numBufferLength == 2:
            self.ignoreLength = 6
        else:
            self.ignoreLength = 7
        self.totalLength += 1

        self.expectedLength = 0
        for i in range(numBufferLength):
            self.expectedLength = (self.expectedLength << 8) \
                + byte_data_list[2 + i]
            self.totalLength += 1

        if self.expectedLength > self.buffer_size:
            self.isPacketOpen = False
            self.lastError = self.INVALID_LENGTH
            return 0

        self.buffer[:self.expectedLength] = byte_data_list[
            2 + numBufferLength:self.expectedLength + 2 + numBufferLength]
        self.totalLength += self.expectedLength

        if byte_data_list[self.expectedLength - 1] != self.frameEnd:
            self.lastError = self.INVALID_LENGTH
            return 0
        self.totalLength += 1

        self.payloadLength = self.totalLength - self.ignoreLength
        print(2 + numBufferLength,
              self.expectedLength + 2 + numBufferLength,
              self.expectedLength, self.payloadLength)
        crc8 = WireCrc()
        crc = crc8.update(self.buffer, self.payloadLength)
        if crc != self.buffer[self.index - 1]:
            self.lastError = self.INVALID_CRC
            return 0

        self.index = 0
        return self.totalLength

    def write_array(self, data):
        return sum(self.write(byte) for byte in data)

    def available(self):
        return 0 if self.isPacketOpen else self.payloadLength - self.index

    def read(self):
        if self.isPacketOpen or self.index >= self.payloadLength:
            return -1
        value = self.buffer[self.index]
        self.index += 1
        return value

    def reset(self):
        self.index = 0
        self.totalLength = 0
        self.expectedLength = 0
        self.payloadLength = 0
        self.isPacketOpen = False
        self.numBufferLength = 0
        self.lastError = self.NONE
        self.ignoreLength = 0
